### DISCUSSION TITLE
Device not appearing on the devices list

### DISCUSSION THREAD

**Discussion by RubenCT98** [dtcooper/raspotify#582] [Q&A] [answered]
Hey, I'm using a Raspberry Pi 3 as my device and I want to run spotify from there on a LAN network. 

I tried following the steps multiples times to install Raspotify and, when I'm done, the device doesn't appear.
I also tried to change the device name and put some credentials on the configuration file at **/etc/raspotify/conf**, but nothing happens.

Checked the service status with **sudo service raspotify status** and that's what I get so far:

> ● raspotify.service - Raspotify (Spotify Connect Client)
>      Loaded: loaded (/lib/systemd/system/raspotify.service; enabled; vendor preset: enabled)
>      Active: active (running) since Wed 2022-10-26 08:53:00 CEST; 27min ago
>        Docs: https://github.com/dtcooper/raspotify
>              https://github.com/librespot-org/librespot
>              https://github.com/dtcooper/raspotify/wiki
>              https://github.com/librespot-org/librespot/wiki/Options
>    Main PID: 1799 (librespot)
>       Tasks: 1 (limit: 1596)
>         CPU: 353ms
>      CGroup: /system.slice/raspotify.service
>              └─1799 /usr/bin/librespot
> 
> d’oct. 26 09:15:52 raspberrypi librespot[1799]: [2022-10-26T07:15:52Z WARN  libmdns::fsm] dropping truncated packet from [fe80::9520:3400:24b0:d868%2]:5353
> d’oct. 26 09:15:52 raspberrypi librespot[1799]: [2022-10-26T07:15:52Z WARN  libmdns::fsm] dropping truncated packet from [fe80::9520:3400:24b0:d868%2]:5353
> d’oct. 26 09:15:52 raspberrypi librespot[1799]: [2022-10-26T07:15:52Z WARN  libmdns::fsm] dropping truncated packet from 192.168.0.52:5353
> d’oct. 26 09:15:52 raspberrypi librespot[1799]: [2022-10-26T07:15:52Z WARN  libmdns::fsm] dropping truncated packet from 192.168.0.52:5353
> d’oct. 26 09:15:52 raspberrypi librespot[1799]: [2022-10-26T07:15:52Z WARN  libmdns::fsm] dropping truncated packet from 192.168.0.52:5353
> d’oct. 26 09:15:52 raspberrypi librespot[1799]: [2022-10-26T07:15:52Z WARN  libmdns::fsm] dropping truncated packet from 192.168.0.52:5353
> d’oct. 26 09:15:53 raspberrypi librespot[1799]: [2022-10-26T07:15:53Z WARN  libmdns::fsm] dropping truncated packet from [fe80::9520:3400:24b0:d868%2]:5353
> d’oct. 26 09:15:53 raspberrypi librespot[1799]: [2022-10-26T07:15:53Z WARN  libmdns::fsm] dropping truncated packet from [fe80::9520:3400:24b0:d868%2]:5353
> d’oct. 26 09:15:53 raspberrypi librespot[1799]: [2022-10-26T07:15:53Z WARN  libmdns::fsm] dropping truncated packet from 192.168.0.52:5353
> d’oct. 26 09:15:53 raspberrypi librespot[1799]: [2022-10-26T07:15:53Z WARN  libmdns::fsm] dropping truncated packet from 192.168.0.52:5353

Not sure what means all of that at the end but the service itself is active and running. Any idea on what's wrong and what I might be missing? I'm completly lost already. I searched for similar problems to mine and multiple resources but none of them seems to have my issue.

**TLDR**: Raspotify installes but device is not appearing on the devices list.

**Comment by JasonLG1979** [comment, marked as answer]
Those are mdns (zeroconf) errors. I'm not a mdns expert by any means but I do know that routers often do not implement mdns correctly, which leads to issues like this. Containers can also have this issue if not properly configured. I can't really debug your router so about your only option is to disable discovery and add your creds to the Raspotify config. The basic setup guide in the wiki has a section about it.

**Reply by RubenCT98** [reply]
I tried to disable discovery, but I'm not sure if I did it properly, since I don't know what kind of parameter should I write to disable it from the config file. I wrote:

`LIBRESPOT_DISABLE_DISCOVERY="true"`

And then, I added my credentials. Restarted the service with:

`sudo systemctl restart raspotify.service`

And after checking the status, this is what I'm getting now:

>  ● raspotify.service - Raspotify (Spotify Connect Client)
>      Loaded: loaded (/lib/systemd/system/raspotify.service; enabled; vendor preset: enabled)
>      Active: active (running) since Thu 2022-10-27 10:00:21 CEST; 1min 11s ago
>        Docs: https://github.com/dtcooper/raspotify
>              https://github.com/librespot-org/librespot
>              https://github.com/dtcooper/raspotify/wiki
>              https://github.com/librespot-org/librespot/wiki/Options
>    Main PID: 7664 (librespot)
>       Tasks: 1 (limit: 1596)
>         CPU: 225ms
>      CGroup: /system.slice/raspotify.service
>              └─7664 /usr/bin/librespot
> 
> d’oct. 27 10:00:21 raspberrypi systemd[1]: Started Raspotify (Spotify Connect Client).
> d’oct. 27 10:00:22 raspberrypi librespot[7664]: [2022-10-27T08:00:22Z WARN  librespot_core::apresolve] Ignoring blacklisted access point ap-gew4.spotify.com:443
> d’oct. 27 10:00:22 raspberrypi librespot[7664]: [2022-10-27T08:00:22Z WARN  librespot_core::apresolve] Ignoring blacklisted access point ap-gue1.spotify.com:80

**EDIT**: These warning messages sometimes changes to only one of them or even change the port of **ap-gue1.spotify.com:80** to **ap-gue.spotify.com:4070**. All that without changing anything, just checking the status after a few minutes.

**Reply by JasonLG1979** [reply]
> I tried to disable discovery, but I'm not sure if I did it properly, since I don't know what kind of parameter should I write to disable it from the config file. I wrote:

> LIBRESPOT_DISABLE_DISCOVERY="true"

It's an option flag. It doesn't matter what you set it to it's mere existence evals to true. So that works.

At this point I'm wondering if it isn't a firewall issue.

**Reply by RubenCT98** [reply]
Right, thanks! At least, now I know it's something of a network issue and not an issue with Raspotify. I'll keep trying to fix that checking at my router, firewall and all that stuff.

Thanks for your time!

**Reply by RubenCT98** [reply]
Just solved some of the issues. My raspberry already appears on the devices list from the Spotify app but now something else happens and I'm not sure if it's still a network issue or maybe some bad configuration:

When I select the raspberry as my device and I press the **Play** button, it changes the device back to the mobile and the raspberry dissappears from the list for a few seconds.

Just in case, I'm leaving here my configuration file:

> 
> \# /etc/raspotify/conf -- Arguments/configuration for librespot
> 
> \# A non-exhaustive list of librespot options and flags.
> 
> \# Please see https://github.com/dtcooper/raspotify/wiki &
> \# https://github.com/librespot-org/librespot/wiki/Options
> \# for configuration details and a full list of options and flags.
> 
> \# You can also find a full list with `librespot -h`.
> 
> \# To avoid name collisions environment variables must be prepended with
> \# `LIBRESPOT_`, so option/flag `foo-bar` becomes `LIBRESPOT_FOO_BAR`.
> 
> \# Invalid environment variables will be ignored.
> 
> \# Raspotify defaults may vary from librespot defaults.
> \# Commenting out the environment variable will fallback to librespot's default
> \# unless otherwise noted.
> 
> \# Flags are either on (uncommented) or off (commented),
> \# their values are otherwise not evaluated (but the "=" is still needed).
> 
> \# Only log warning and error messages.
> LIBRESPOT_QUIET=
> 
> \# Automatically play similar songs when your music ends.
> LIBRESPOT_AUTOPLAY=
> 
> \# Disable caching of the audio data.
> \# Enabling audio data caching can take up a lot of space
> \# if you don't limit the cache size with LIBRESPOT_CACHE_SIZE_LIMIT.
> \# It can also wear out your Micro SD card. You have been warned. 
> LIBRESPOT_DISABLE_AUDIO_CACHE=
> 
> \# Disable caching of credentials.
> \# Caching of credentials is not necessary so long as
> \# LIBRESPOT_DISABLE_DISCOVERY is not set.
> LIBRESPOT_DISABLE_CREDENTIAL_CACHE=
> 
> \# Play all tracks at approximately the same apparent volume.
> LIBRESPOT_ENABLE_VOLUME_NORMALISATION=
> 
> \# Enable verbose log output.
> \#LIBRESPOT_VERBOSE=
> 
> \# Disable zeroconf discovery mode.
> LIBRESPOT_DISABLE_DISCOVERY=
> 
> \# Options will fallback to their defaults if commented out,
> \# otherwise they must have a valid value.
> 
> \# Device name.
> \# Raspotify defaults to "raspotify (*hostname)".
> \# Librespot defaults to "Librespot".
> \#LIBRESPOT_NAME="Librespot"
> 
> \# Bitrate (kbps) {96|160|320}. Defaults to 160.
> LIBRESPOT_BITRATE="320"
> 
> \# Output format {F64|F32|S32|S24|S24_3|S16}. Defaults to S16.
> \#LIBRESPOT_FORMAT="S16"
> 
> \# Displayed device type. Defaults to speaker.
> \#LIBRESPOT_DEVICE_TYPE="speaker"
> 
> \# Limits the size of the cache for audio files.
> \# It's possible to use suffixes like K, M or G, e.g. 16G for example.
> \# Highly advised if audio caching isn't disabled. Otherwise the cache
> \# size is only limited by disk space.
> \#LIBRESPOT_CACHE_SIZE_LIMIT=""
> 
> \# Audio backend to use, alsa or pulseaudio. Defaults to alsa.
> \#LIBRESPOT_BACKEND="alsa"
> 
> \# Username used to sign in with.
> \# Credentials are not required if LIBRESPOT_DISABLE_DISCOVERY is not set.
> LIBRESPOT_USERNAME="xxxxx"
> 
> \# Password used to sign in with.
> LIBRESPOT_PASSWORD="xxxxx"
> 
> \# Audio device to use, use `librespot --device ?` to list options.
> \# Defaults to the system's default.
> \#LIBRESPOT_DEVICE="default"
> 
> \# Initial volume in % from 0 - 100.
> \# Defaults to 50 For the alsa mixer: the current volume.
> \#LIBRESPOT_INITIAL_VOLUME="50"
> 
> \# Volume control scale type {cubic|fixed|linear|log}.
> \# Defaults to log.
> \#LIBRESPOT_VOLUME_CTRL="log"
> 
> \# Range of the volume control (dB) from 0.0 to 100.0.
> \# Default for softvol: 60.0.
> \# For the alsa mixer: what the control supports.
> \#LIBRESPOT_VOLUME_RANGE="60.0"
> 
> \# Pregain (dB) applied by volume normalisation from -10.0 to 10.0.
> \# Defaults to 0.0.
> \#LIBRESPOT_NORMALISATION_PREGAIN="0.0"
> 
> \# Threshold (dBFS) at which point the dynamic limiter engages
> \# to prevent clipping from 0.0 to -10.0.
> \# Defaults to -2.0.
> \#LIBRESPOT_NORMALISATION_THRESHOLD="-2.0"
> 
> \# The port the internal server advertises over zeroconf 1 - 65535.
> \# Ports <= 1024 may require root privileges.
> \#LIBRESPOT_ZEROCONF_PORT=""
> 
> \# HTTP proxy to use when connecting.
> \#LIBRESPOT_PROXY=""
> 
> \# ### This is NOT a librespot option or flag. ###
> \# This modifies the behavior of the Raspotify service.
> \# If you have issues with this option DO NOT file a bug with librespot.
> \# 
> \# By default librespot "download buffers" tracks, meaning that it downloads
> \# the tracks to disk and plays them from the disk and then deletes them when
> \# the track is over. This practice is very common, many other audio frameworks
> \# and players do the exact same thing as a disk based tmp cache is easy to use
> \# and very resilient. That being said there may be cases where a user may want
> \# to minimize disk read/writes.
> \#
> \# Commenting this out will cause librespot to use a tmpfs so that provided there
> \# is enough RAM to hold the track nothing is written to disk but instead to a tmpfs.
> \# See https://github.com/dtcooper/raspotify/discussions/567
> \# And https://www.kernel.org/doc/html/latest/filesystems/tmpfs.html
> TMPDIR=/tmp

**Reply by JasonLG1979** [reply]
> When I select the raspberry as my device and I press the Play button, it changes the device back to the mobile and the raspberry dissappears from the list for a few seconds.

Do the logs say anything?

If it's disappearing when you try to play something and then reappearing that sounds a lot like it's crashing and being restarted by systemd.

**Reply by RubenCT98** [reply]
That's what I got from doing `journalctl | grep raspotify`:

> d’oct. 31 08:49:09 raspberrypi systemd[1]: raspotify.service: Main process exited, code=exited, status=1/FAILURE
> d’oct. 31 08:49:09 raspberrypi systemd[1]: raspotify.service: Failed with result 'exit-code'.
> d’oct. 31 08:49:09 raspberrypi systemd[1]: raspotify.service: Consumed 1.700s CPU time.
> d’oct. 31 08:49:19 raspberrypi systemd[1]: raspotify.service: Scheduled restart job, restart counter is at 10.
> d’oct. 31 08:49:19 raspberrypi systemd[1]: raspotify.service: Consumed 1.700s CPU time.

Not sure if there're other logs to check for Raspotify or if that's all, but yeah, seems like it's crashing and being restarted by systemd as you said.

**Reply by RubenCT98** [reply]
Small update:

I changed the command to read the journalctl logs and I found some other info. I did `journalctl --system _SYSTEMD_UNIT=raspotify.service` and now I'm getting this when the crash happens:

> de nov. 01 23:28:21 raspberrypi librespot[1657]: Failed to get home directory.
> de nov. 01 23:28:21 raspberrypi librespot[1657]: Failed to get home directory.
> de nov. 01 23:28:21 raspberrypi librespot[1657]: Failed to load cookie file from cookie: No such file or directory
> de nov. 01 23:28:21 raspberrypi librespot[1657]: Failed to get home directory.
> de nov. 01 23:28:21 raspberrypi librespot[1657]: Failed to get home directory.
> de nov. 01 23:28:21 raspberrypi librespot[1657]: Failed to load cookie file from cookie: No such file or directory
> de nov. 01 23:28:21 raspberrypi librespot[1657]: ALSA lib pulse.c:242:(pulse_connect) PulseAudio: Unable to connect: Access denied
> de nov. 01 23:28:21 raspberrypi librespot[1657]: [2022-11-01T22:28:21Z ERROR librespot_playback::player] Audio Sink Error Connection Refused: <AlsaSink> Device default May be Invalid, >
> de nov. 01 23:28:32 raspberrypi librespot[1666]: [2022-11-01T22:28:32Z WARN  librespot_core::apresolve] Ignoring blacklisted access point ap-gew4.spotify.com:4070

**Reply by JasonLG1979** [reply]
The read me clearly states that Raspotify is designed to be ran on headless systems.

**Reply by RubenCT98** [reply]
Yes, it's running on a headless system. Exactly on the latest Raspbian OS. I'm doing all the configuration from an external computer, connecting from SSH, and when I test if it works, I do it from a mobile.

**Reply by JasonLG1979** [reply]
A headless system with pulsesudio installed?

**Reply by RubenCT98** [reply]
I was using alsa at first, then tried with pulseaudio, just in case there was something wrong with it. I thought on pulseaudio because on the config file, the LIBRESPOT_BACKEND parameter asks for alsa or pulseaudio, using alsa as default, so I thought on trying to use pulseaudio from that.

**Reply by JasonLG1979** [reply]
The pulsesudio backend is included for advanced users and is very much "batteries not included" and not supported.

**Reply by JasonLG1979** [reply]
Basically if you want my help don't run Raspotify on a system with pulsesudio installed.

**Comment by jackobtone** [comment]
I will hook up to this topic. Recently I`ve installed DietPi + Alsa + Raspotify. Everything worked till todays update (got a prompt in the system). After it i cant find Raspberry in Spotify app. I have reinstalled the system, installed Alsa, installed Raspotify just like I did before and it didnt work. Command _sudo service raspotify status_ give back a message:

> root@DietPi:~# sudo service raspotify status
● raspotify.service - Raspotify (Spotify Connect Client)
     Loaded: loaded (/lib/systemd/system/raspotify.service; disabled; vendor preset: enabled)
     Active: active (running) since Mon 2022-12-05 18:15:42 GMT; 32s ago
       Docs: https://github.com/dtcooper/raspotify
             https://github.com/librespot-org/librespot
             https://github.com/dtcooper/raspotify/wiki
             https://github.com/librespot-org/librespot/wiki/Options
    Process: 1223 ExecStartPre=/bin/sleep 10 (code=exited, status=0/SUCCESS)
   Main PID: 1224 (librespot)
      Tasks: 1 (limit: 4531)
        CPU: 142ms
     CGroup: /system.slice/raspotify.service
             └─1224 /usr/bin/librespot
Dec 05 18:15:32 DietPi systemd[1]: Starting Raspotify (Spotify Connect Client)...
Dec 05 18:15:42 DietPi systemd[1]: Started Raspotify (Spotify Connect Client).

I assume everything should work, but it doesnt

**Reply by JasonLG1979** [reply]
Open a proper bug report.
